src/backend/OpenCL/CHIPBackendOpenCL.hh

```diff
@@ -86,30 +84,46 @@ public:
 };
 
 class CHIPEventOpenCL : public CHIPEvent {
-public:
-  cl_event ClEvent;
-  friend class CHIPEventOpenCL;
+private:
+  cl::Event ClEvent;
 
 public:
   CHIPEventOpenCL(CHIPContextOpenCL *ChipContext, cl_event ClEvent,
                   CHIPEventFlags Flags = CHIPEventFlags(),
                   bool UserEvent = false);
   CHIPEventOpenCL(CHIPContextOpenCL *ChipContext,
                   CHIPEventFlags Flags = CHIPEventFlags());
-  virtual ~CHIPEventOpenCL() override;
+  virtual ~CHIPEventOpenCL() override{};
   virtual void recordStream(CHIPQueue *ChipQueue) override;
-  void takeOver(CHIPEvent *Other);
+  virtual size_t getCHIPRefc() override;
   bool wait() override;
   float getElapsedTime(CHIPEvent *Other) override;
-  virtual void hostSignal() override;
+
+  // not needed anywhere
+  virtual void hostSignal() override{};
+
   virtual bool updateFinishStatus(bool ThrowErrorIfNotReady = true) override;
-  cl_event *getNativePtr() { return &ClEvent; }
-  cl_event &getNativeRef() { return ClEvent; }
-  uint64_t getFinishTime();
-  size_t getRefCount();
+  cl::Event &get() { return ClEvent; }
+  cl::UserEvent &getAsUserEv() { return static_cast<cl::UserEvent &>(ClEvent); }
+  void reset(cl::UserEvent &&Ev) {
+    ClEvent = Ev;
+    logTrace("UserEvent {} Moved into {} || NOW: {}", (void *)Ev.get(),
+             (void *)this, (void *)ClEvent.get());
+  }
 
-  virtual void increaseRefCount(std::string Reason) override;
-  virtual void decreaseRefCount(std::string Reason) override;
+  void reset(cl::Event &&Ev) {
+    ClEvent = Ev;
+    logTrace("Event {} Moved into {} || NOW: {}", (void *)Ev.get(),
+             (void *)this, (void *)ClEvent.get());
+  }
+  void reset(cl_event Ev) {
+    ClEvent = Ev;
+    logTrace("Event {} Moved into {} ||| NOW: {}", (void *)Ev, (void *)this,
+             (void *)ClEvent.get());
+  }
```

Comment:
Why does event reset take any arguments at all, especially a new event? Should be `init()` or something similar.

Reply:
This is meant to mimic `std::shared_ptr<T>::reset()` - it replaces the object (ClEvent) stored currently, and releases the old one. `init()` is usually for stuff that needs to be done once, but for some reason can't go into the constructor, so that's quite different.